de-lib/src/main/java/org/iplantc/de/apps/client/presenter/hierarchies/OntologyHierarchiesPresenterImpl.java

```diff
@@ -105,6 +106,58 @@ public void onSuccess(final App appDetails) {
         }
     }
 
+    class FilteredHierarchyCallback implements AsyncCallback<OntologyHierarchy> {
+        private final Tree<OntologyHierarchy, String> tree;
+        private final OntologyHierarchy root;
+        private final OntologyHierarchy selectedHierarchy;
+        private final int rootsCount;
+        private int callbackCount;
+        private boolean lastCallback = false;
+
+        public FilteredHierarchyCallback(Tree<OntologyHierarchy, String> tree,
+                                         OntologyHierarchy root,
+                                         OntologyHierarchy selectedHierarchy,
+                                         int rootsCount) {
+            this.tree = tree;
+            this.root = root;
+            this.selectedHierarchy = selectedHierarchy;
+            this.rootsCount = rootsCount;
+            this.callbackCount = 0;
+        }
+
+        @Override
+        public void onFailure(Throwable caught) {
+            ErrorHandler.post(caught);
+            tree.unmask();
+        }
+
+        @Override
+        public void onSuccess(OntologyHierarchy result) {
+            if (result == null || result.getSubclasses() == null) {
+                result = root;
+                result.setSubclasses(Lists.<OntologyHierarchy>newArrayList());
+            }
+            OntologyHierarchy unclassifiedChild = ontologyUtil.addUnclassifiedChild(result);
+            unclassifiedHierarchies.add(unclassifiedChild);
+            //Set the key for the current root (which won't appear in the tree, but will be the name of the tab)
+            // which will allow the children to know the full path from its parent to node
+            ontologyUtil.getOrCreateHierarchyPathTag(result);
+            addHierarchies(tree.getStore(), null, result.getSubclasses());
+            checkForDesiredHierarchy();
+
+            tree.unmask();
+        }
+
+        void checkForDesiredHierarchy() {
+            callbackCount++;
```

Comment:
I'd like some eyes on this portion.  I'm not sure if this is the best way to keep track of whether the hierarchy the user was previously viewing is still in the ontology after they've Refreshed.
I'm basically keeping a counter that increments per callback. Once I've reached the last callback's return, I check to see if the hierarchy was ever found. If not, then I fire an event telling `AppCategoryPresenterImpl` to go ahead and select/load the `Apps under Development` category instead.  Since callbacks are asynchronous I'm not sure if I'm missing anything.